zeth-contracts/contracts/BaseMixer.sol

```diff
@@ -98,20 +126,121 @@ contract BaseMixer is MerkleTreeMiMC7, ERC223ReceivingContract {
+        vpub_in = uint64(Bytes.get_last_8_bytes(uint(vpub_bytes)));
+
+        // We retrieve the public value out, remove any extra bits (due to the padding) and inverse the bit order
+        vpub_bytes = bytes32(primary_inputs[1 + nb_hash_digests]) >> residual_hash_size;
+        vpub_out = uint64(Bytes.get_last_8_bytes(uint(vpub_bytes)));
+    }
+
+    // This function is used to reassemble hsig given the the primary_inputs
+    // To do so, we extract the remaining bits of hsig from the residual field element(S) and combine them with the hsig field element
+    function assemble_hsig(uint[] memory primary_inputs) public pure returns (bytes32 hsig){
+
+        // We know hsig residual bits correspond to the 128th to 130st bits of the first residual field element after padding.
+        // We retrieve hsig's residual bits and remove any extra bits (due to the padding)
+        bytes32 hsig_bytes = (bytes32(primary_inputs[1 + nb_hash_digests]) << padding_size + 2*public_value_length) >> field_capacity;
+        bytes1 bits_input = Bytes.get_last_byte(hsig_bytes);
+
+        // We reassemble the residual bits with the field element
+        hsig = Bytes.sha256_digest_from_field_elements(primary_inputs[1 + jsIn + jsOut] << (digest_length - field_capacity), bits_input);
+    }
+
+    // This function is used to reassemble the nullifiers given the nullifier index [0, jsIn[ and the primary_inputs
+    // To do so, we extract the remaining bits of the nullifier from the residual field element(S) and combine them with the nullifier field element
+    function assemble_nullifier(uint index, uint[] memory primary_inputs) public pure returns (bytes32 sn){
+        // We first check that the nullifier we want to retrieve exists
+        require(
+            index < jsIn,
+            "nullifier index overflow"
+        );
+        // We offset the nullifier index by the number of values preceding the nullifiers in the primary inputs:
+        // the root (1)
+        uint nullifier_index = 1 + index;
+
+        // We compute the nullifier's residual bits index and check the 1st f.e. indeed comprises it
+        // See the way the residual bits are ordered in the extended proof
+        uint sn_bit_index = 2*public_value_length + nullifier_index * packing_residue_length;
+        require(
+            field_capacity >= sn_bit_index + packing_residue_length,
+            "nullifier written in different residual bit f.e."
+        );
+
+        // We retrieve nf's residual bits and remove any extra bits (due to the padding)
+        bytes32 sn_bytes = (bytes32(primary_inputs[1 + nb_hash_digests]) << padding_size + sn_bit_index) >> field_capacity;
+        bytes1 bits_input = Bytes.get_last_byte(sn_bytes);
+        // We reassemble the residual bits with the field element
+        sn = Bytes.sha256_digest_from_field_elements(primary_inputs[nullifier_index] << (digest_length - field_capacity), bits_input);
```

Comment:
`sn` should be `nf` for consistency. We should stick to `nullifiers`

Reply:
Fixed.